pkg/operator/status.go

```diff
@@ -289,9 +301,165 @@ func (optr *Operator) syncUpgradeableStatus() error {
+			coStatus.Message = status.Message
+			mcoObjectRef := &corev1.ObjectReference{
+				Kind:      co.Kind,
+				Name:      co.Name,
+				Namespace: co.Namespace,
+				UID:       co.GetUID(),
+			}
+			optr.eventRecorder.Eventf(mcoObjectRef, corev1.EventTypeWarning, coStatus.Reason, fmt.Sprintf("machine config pool contains nodes with an unsupported kubelet version skew."))
+			return optr.updateStatus(co, coStatus)
+		case skewPresent:
+			coStatus.Reason = status.Reason
+			coStatus.Message = status.Message
+			mcoObjectRef := &corev1.ObjectReference{
+				Kind:      co.Kind,
+				Name:      co.Name,
+				Namespace: co.Namespace,
+				UID:       co.GetUID(),
+			}
+			optr.eventRecorder.Eventf(mcoObjectRef, corev1.EventTypeWarning, coStatus.Reason, fmt.Sprintf("machine config pool contains nodes with an N-1 kubelet version skew."))
+			return optr.updateStatus(co, coStatus)
+		}
+	}
 	return optr.updateStatus(co, coStatus)
 }
 
+// isKubeletSkewSupported checks the version skew of kube-apiserver and node kubelet version.
+// Returns the versionSkew. version skew > 2 is not supported.
+func (optr *Operator) isKubeletSkewSupported() (skew versionSkew, coStatus configv1.ClusterOperatorStatusCondition, err error) {
+	coStatus = configv1.ClusterOperatorStatusCondition{}
+	kubeAPIServerStatus, err := optr.configClient.ConfigV1().ClusterOperators().Get(context.TODO(), "kube-apiserver", metav1.GetOptions{})
+	if err != nil {
+		coStatus.Reason = "KubeletSkewUnchecked"
+		coStatus.Message = fmt.Sprintf("An error occurred when checking kubelet version skew: %v", err)
+		return skewUnchecked, coStatus, err
+	}
+	// looks like
+	//   - name: kube-apiserver
+	//     version: 1.21.0-rc.0
+	kubeAPIServerVersion := ""
+	for _, version := range kubeAPIServerStatus.Status.Versions {
+		if version.Name != "kube-apiserver" {
+			continue
+		}
+		kubeAPIServerVersion = version.Version
+		break
+	}
+	if kubeAPIServerVersion == "" {
+		err = fmt.Errorf("kube-apiserver does not yet have a version")
+		coStatus.Reason = "KubeletSkewUnchecked"
+		coStatus.Message = fmt.Sprintf("An error occurred when checking kubelet version skew: %v", err.Error())
+		return skewUnchecked, coStatus, err
+	}
+	kubeAPIServerMinorVersion, err := getMinorKubeletVersion(kubeAPIServerVersion)
+	if err != nil {
+		coStatus.Reason = "KubeletSkewUnchecked"
+		coStatus.Message = fmt.Sprintf("An error occurred when checking kubelet version skew: %v", err)
+		return skewUnchecked, coStatus, err
+	}
+	var (
+		lastError      error
+		kubeletVersion string
+		nodes          []*corev1.Node
+	)
+	masterNodes, err := optr.nodeLister.List(labels.SelectorFromSet(labels.Set{ctrlcommon.MasterLabel: ""}))
+	if err != nil {
+		coStatus.Reason = "KubeletSkewUnchecked"
+		coStatus.Message = fmt.Sprintf("An error occurred when checking kubelet version skew: %v", err)
+		return skewUnchecked, coStatus, err
+	}
+	workerNodes, err := optr.nodeLister.List(labels.SelectorFromSet(labels.Set{node.WorkerLabel: ""}))
```

Comment:
Clusters can have custom pools as well.

Reply:
They can have custom pools, but they are always Masters or Workers, right?

Reply:
custom pools don't have to have a worker label: https://github.com/openshift/machine-config-operator/blob/master/docs/custom-pools.md#creating-a-custom-pool

Reply:
Does those nodes have label llike `labels.SelectorFromSet(labels.Set{fmt.Sprintf("node-role.kubernetes.io/%s", pool.Name): ""})`?
If they do we can list the pools and filter the nodes by the label.

Reply:
@kikisdeliveryservice  As https://github.com/openshift/machine-config-operator/pull/2658#issuecomment-885218200 commeted, the new function will get the mco managed node.